app/controllers/courses_controller.rb

```diff
@@ -0,0 +1,250 @@
+    def drafts
+        c_user = current_user()
+        return ( redirect_to :root ) unless logged_in?  # Ensure the user is logged in
+
+        courses = Course.where(visibility: Visibility.draft, user_id: c_user.id).order(updated_at: :desc)
+        render :json => { "items": courses }
+    end
+
+    # Request: GET /courses/published
+    # Authorization: Available to public
+    # Unauthorized access: Should get redirected to root (/)
+    # Authorized access: Return a JSON of all courses where course.visibility == Visibility.published in the format { items: [ course1, course2, ...] }
+    def published        
+        courses = Course.where(visibility: Visibility.published).order(updated_at: :desc)
+        render :json => { "items": courses }
+    end
+
+    # Request: GET /courses/(:course_id)
+    # Authorization:
+    # - If course.visibility == Visibilityi.publisehd then visible to public
+    # - otherwise only available to admin, moderator, and creator of course
+    # Unauthorized access: Should get redirected to root (/)
+    # Authorized access: Return a JSON of the course where course.id == course_id in the format { item: { title: "How to kill a mockingbird", description : "...", ... } }
+    def getcourse
+        c_user = current_user()
+        return ( redirect_to :root ) unless logged_in?  # Ensure the user is logged in
+        course = Course.where(id: params[:course_id]).first()
+
+        # Course not found case
+        if course.nil?
+            return ( render :file => "#{Rails.root}/public/404.html", :status => 404 )
+        end
+
+        # Public case:
+        if course.visibility == Visibility.published
+            render :json => course
+
+        # Authenticated case:
+        elsif logged_in?  # Ensure that the user is logged in
+            if c_user.role == Role.admin or 
+                    c_user.role == Role.moderator or
+                    c_user.id == course.user_id then
+                render :json => course
+            else
+                redirect_to :root
+            end
+
+        # Otherwise If protected course and not logged in:
+        else
+            redirect_to :root
+        end
+    end
+
+    # Request: GET /courses/new
+    # Authorization: Only available to logged in users
+    # Unauthorized access: Should get redirected to root (/)
+    # Authorized access: Create a new course in database based on template.
+    # Return a JSON form for editing the new course. Should contain:
+    # - A Rails-generated CSRF token for transactions
+    # - The template course with all its fields and fill in any fields with the appropriate existing values
+    def new
+        c_user = current_user()
+        return ( redirect_to :root ) unless logged_in?  # Ensure the user is logged in
+        token = session[:_csrf_token]
+
+        course = Course.create(user_id: c_user.id)
```

Comment:
I'm not sure what is happening in this one. The course should not be persisting to database until the post comes back with a filled form, and the save/redir below shouldn't be there either. This a GET request, so by rule we have to make it non-modifying to the database.

The goal is to just provide a template/form for the client to fill out.

Reply:
So it should be an unconditional `render :json => { "token": token, "course": course }` after this